src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -6025,7 +6025,12 @@ public static Task WhenAll(params Task[] tasks)
         /// </remarks>
         /// <exception cref="ArgumentException">The <paramref name="tasks"/> array contained a null task.</exception>
         public static Task WhenAll(params ReadOnlySpan<Task> tasks) =>
-            tasks.Length != 0 ? new WhenAllPromise(tasks) : CompletedTask;
+            tasks.Length switch
+            {
+                0 => Task.CompletedTask,
+                1 => tasks[0],
```

Comment:
This changes behavior if task[0] is null. It needs to be null checked and trigger the same exception as occurs in WhenAllPromise's ctor.